packages/from-hardhat/src/compilation.ts

```diff
@@ -0,0 +1,121 @@
+import type * as Hardhat from "hardhat/types";
+
+import type { Compilation, CompiledContract } from "@truffle/compile-common";
+import * as CompileSolidity from "@truffle/compile-solidity";
+
+import { supportedHardhatBuildInfoFormats } from "./constants";
+
+export const fromBuildInfo = (buildInfo: Hardhat.BuildInfo): Compilation => {
+  const { _format } = buildInfo;
+
+  if (!supportedHardhatBuildInfoFormats.has(_format)) {
+    throw new Error(`Unsupported build info format: ${_format}`);
+  }
+
+  const sourceIndexes = SourceIndexes.fromBuildInfo(buildInfo);
+
+  return {
+    sourceIndexes,
+    sources: Sources.fromBuildInfo(buildInfo, sourceIndexes),
+    contracts: Contracts.fromBuildInfo(buildInfo),
+    compiler: {
+      name: "solc",
+      version: buildInfo.solcLongVersion
+    }
+  };
+};
+
+namespace SourceIndexes {
+  export const fromBuildInfo = (
+    buildInfo: Hardhat.BuildInfo
+  ): Compilation["sourceIndexes"] => {
+    const sourceIndexes = [];
+    for (const { index, sourcePath } of Object.entries(
+      buildInfo.output.sources
+    ).map(([sourcePath, source]) => ({ index: source.id, sourcePath }))) {
+      sourceIndexes[index] = sourcePath;
+    }
```

Comment:
Same question about sparse arrays here as well

Reply:
Yeah I should just remove the sparseness support.

(I added it because @truffle/db allows everything to be sparse, and I had forgotten that that wasn't true of compile-common's spec)